### ISSUE TITLE
X11: DeviceEvent::Key modifiers never set

### ISSUE THREAD

**Issue by gmorenz** [rust-windowing/winit#483]
Arch Linux  
x11
winit - all tested: 0.11.3 / 0.12.0 / current git head (eadd9a19b27ab3f6aa4904c5cb31726c5591cfd9) 

All the fields of `ModifierState` are always false, regardless of whether or not I am holding shift/ctrl/alt/windows key. Shift/ctrl/false/window key presses and releases are registered as key event's and passed to my application, they just don't change the modifier state.

Test code:

```rust
extern crate winit;

use winit::{Event, EventsLoop, DeviceEvent, WindowEvent, Window, ControlFlow};

fn main() {
    let mut events_loop = EventsLoop::new();
    let _window = Window::new(&events_loop);

    events_loop.run_forever(|event| {
        match event {
            Event::WindowEvent { event: WindowEvent::CloseRequested, .. } => {
                return ControlFlow::Break;
            },
            Event::DeviceEvent { event: DeviceEvent::Key(key), .. } => {
                println!("{:?}", key);
                if key.modifiers.shift == true {
                    panic!("... I can't get here?");
                }
            },
            _ => ()
        };

        ControlFlow::Continue
    });
}
```

**Comment by gmorenz** [comment]
Ah, you're right, `WindowEvent` works correctly, I'm guessing this is intentional and counts as user error then?

**Comment by francesca64** [comment]
Well, *intentional* is a strong word. It's just that the relevant X11 event has no modifier info associated with it, and so the only straightforward solution for this is to make a round-trip to the X server (which is something I've been working on reducing).

I've thought of a cool idea for how to fix this, so I'll reopen this.

**Comment by xla** [comment]
@francesca64 Ran into this issue as well, Is it advised to track the state of the modifier keys independently? I'd like to perform an action when the user presses e.g. `alt + n`.

**Comment by francesca64** [comment]
@xla I'm not active here anymore. [The X11 backend doesn't currently have a maintainer](https://github.com/rust-windowing/winit/blob/b94572621a4da21af909950d8a00211b053c4d5f/CONTRIBUTING.md#maintainers--testers), so this bug probably won't be fixed for a while. You only need this feature if you want to handle modifiers while the window isn't focused; @Osspial is a better person to ask if you need more advice.

For anyone who wants to fix this in the future, using the Xkb extension is the only way. This is esoteric, but I have a reference implementation here: https://github.com/francesca64/winit/commit/fe0e98e72733616ebfbb54ba7462e0d55f6f55ab (I remember there being another branch from my second attempt to do this, but I didn't see one). The normal way to interpret the Xkb state info is using libxkbcommon, which is ubiquitous, *but* libxkbcommon-x11 isn't! So, to implement this, parts of libxkbcommon would have to be reimplemented.

**Comment by murarth** [comment]
Closed by #1132, which implemented manual modifier state tracking in X11.
